Use newer AWS API for paginated queries

bboreham opened the pull request:
**What this PR does**:

Call `QueryPagesWithContext()` instead of iterating through `NextPage()`.
Use AWS-SDK `Handlers` for tracing each retry and reporting errors - trace output will be different with just one span for the whole query and span-log entries for each page and retried operation.

This is less code, and more robust when retrying requests.  I strongly suspect that before this change it could get into a state on network errors where it would fail every time until the max number of retrys was hit - error message looks like:
```
level=warn ts=2020-04-10T19:30:30Z msg="DynamoDB error" retry=19 table=prod_chunk_index2_2565 err="RequestError: send request failed\ncaused by: Post https://dynamodb.us-east-1.amazonaws.com/: net/http: HTTP/1.x transport connection broken: http: ContentLength=205 with Body length 0"
```

We don't need so much from the request object for testing now.

There are still two operations where we use `request.Send()`, but they do not use pagination so no benefit to rewrite in a similar way.


**Which issue(s) this PR fixes**:
Fixes #403
Part of #1152 

**Checklist**
- [x] Tests updated
- NA Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

gouthamve commented on pkg/chunk/aws/dynamodb_storage_client.go:
```diff
@@ -327,88 +326,44 @@ func (a dynamoDBStorageClient) query(ctx context.Context, query chunk.IndexQuery
-	var err error
-	for backoff.Ongoing() {
-		err = instrument.CollectedRequest(ctx, "DynamoDB.QueryPages", dynamoRequestDuration, instrument.ErrorCode, func(innerCtx context.Context) error {
+		return a.DynamoDB.QueryPagesWithContext(ctx, input, func(output *dynamodb.QueryOutput, _ bool) bool {
```
Shouldn't we use `innerCtx` here?

pracucci commented on pkg/chunk/aws/retryer.go:
```diff
@@ -0,0 +1,53 @@
+	}
+}
+
+func (r *retryer) withRetrys(req *request.Request) {
```
Did you mean `withRetries()`?

bboreham commented on pkg/chunk/aws/retryer.go:
```diff
@@ -0,0 +1,53 @@
+
+// ShouldRetry returns if the failed request is retryable.
+func (r *retryer) ShouldRetry(req *request.Request) bool {
+	var d client.DefaultRetryer
```
Thanks - hadn't spotted that.  Now I look back at this code, I was trying to re-use the AWS-SDK behaviour, but it's probably clearer just to copy-paste it.

bboreham commented on pkg/chunk/aws/dynamodb_storage_client.go:
```diff
@@ -327,88 +326,44 @@ func (a dynamoDBStorageClient) query(ctx context.Context, query chunk.IndexQuery
+		}, retryer.withRetrys, withErrorHandler(query.TableName, "DynamoDB.QueryPages"))
+	})
+	if err != nil {
+		return fmt.Errorf("QueryPage error: table=%v, err=%v", query.TableName, err)
```
Maybe that would be better, but note I didn't change this error in this PR; that's what it did before.

bboreham commented:
I believe I've fixed all the points made.
Regarding retries=0, it doesn't seem to be documented to do infinite retries; I suspect that came out of a merger between two different bits of retry code.  

I have also added a check that the context hasn't been cancelled, via `Backoff.Ongoing()`.

bboreham commented:
Rebased.
